File: api/Portfolio.Api/Services/ResumeTools.cs
using System.Text.Json;
using Portfolio.Api.Models;

namespace Portfolio.Api.Services;

/// <summary>
/// Pure handlers backing the chat agent's tool calls. Each takes the model's
/// JSON input, validates against <see cref="ResumeToolInputValidator"/>, then
/// queries <see cref="ResumeDataService"/> and returns JSON the model can cite.
/// </summary>
public sealed class ResumeTools(ResumeDataService resumeDataService, TimeProvider timeProvider)
{
    private readonly ResumeData _resume = resumeDataService.Data;
    private readonly TimeProvider _time = timeProvider;

    /// <summary>
    /// Dispatch by tool name. Unknown names and validation failures yield structured
    /// <c>{ "error": ... }</c> payloads so the streaming loop can emit recoverable tool_result rows.
    /// </summary>
    public Task<JsonElement> RunAsync(string name, JsonElement input, CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();

        if (!ResumeToolInputValidator.TryValidate(name, input, out var schemaError))
            return Task.FromResult(ValidationError(schemaError!));

        JsonElement result = name switch
        {
            ResumeToolDefinitions.GetRole => RunGetRole(input),
            ResumeToolDefinitions.SearchResume => RunSearchResume(input),
            ResumeToolDefinitions.ListProjectsBySkill => RunListProjectsBySkill(input),
            ResumeToolDefinitions.GetMetrics => RunGetMetrics(input),
            ResumeToolDefinitions.ListRecentShipped => RunListRecentShipped(input),
            ResumeToolDefinitions.GetNarrative => RunGetNarrative(),
            ResumeToolDefinitions.GetFaq => RunGetFaq(input),
            _ => Error($"Unknown tool '{name}'."),
        };

        return Task.FromResult(result);
    }

    private JsonElement RunGetRole(JsonElement input)
    {
        if (!ResumeToolInputValidator.TryDeserializeGetRole(input, out var model, out var err) || model is null)
            return ValidationError(err ?? "deserialize_failed");

        var currentYear = _time.GetUtcNow().Year;
        var matches = _resume.Roles
            .Where(r => MatchesRole(r, model.Id, model.Org, model.Year, currentYear))
            .Select(RoleToObject)
            .ToList();

        return JsonSerializer.SerializeToElement(new { items = matches, count = matches.Count });
    }

    private static bool MatchesRole(ResumeRole role, string? id, string? org, int? year, int currentYear)
    {
        if (id is not null && !string.Equals(role.Id, id, StringComparison.OrdinalIgnoreCase))
            return false;

        if (org is not null && role.Org.IndexOf(org, StringComparison.OrdinalIgnoreCase) < 0)
            return false;

        if (year is { } y)
        {
            var endYear = role.EndYear ?? currentYear;
            if (y < role.StartYear || y > endYear) return false;
        }

        return true;
    }

    private JsonElement RunSearchResume(JsonElement input)
    {
        if (!ResumeToolInputValidator.TryDeserializeSearchResume(input, out var model, out var err) || model is null)
            return ValidationError(err ?? "deserialize_failed");

        if (string.IsNullOrWhiteSpace(model.Query))
            return Error("'query' is required.");

        var query = model.Query.Trim();
        var items = new List<object>();

        foreach (var r in _resume.Roles)
        {
            var matched = MatchRoleFields(r, query);
            if (matched.Count > 0)
                items.Add(new { kind = "role", id = r.Id, matchedFields = matched });
        }

        foreach (var p in _resume.Projects)
        {
            var matched = MatchProjectFields(p, query);
            if (matched.Count > 0)
                items.Add(new { kind = "project", id = p.Id, matchedFields = matched });
        }

        var narrativeMatched = MatchNarrativeFields(_resume.Narrative, query);
        if (narrativeMatched.Count > 0)
            items.Add(new { kind = "narrative", id = "narrative", matchedFields = narrativeMatched });

        foreach (var f in _resume.Faq)
        {
            var matched = MatchFaqFields(f, query);
            if (matched.Count > 0)
                items.Add(new { kind = "faq", id = f.Id, matchedFields = matched });
        }

        var personMatched = MatchPersonFields(_resume.Person, query);
        if (personMatched.Count > 0)
            items.Add(new { kind = "person", id = "person", matchedFields = personMatched });

        return JsonSerializer.SerializeToElement(new { query, items, count = items.Count });
    }

    private static List<string> MatchRoleFields(ResumeRole r, string query)
    {
        var fields = new List<string>();
        if (ContainsIgnoreCase(r.Title, query)) fields.Add("title");
        if (ContainsIgnoreCase(r.Org, query)) fields.Add("org");
        if (ContainsIgnoreCase(r.Summary, query)) fields.Add("summary");
        if (r.Achievements.Any(a => ContainsIgnoreCase(a, query))) fields.Add("achievements");
        if (r.Tech.Any(t => ContainsIgnoreCase(t, query))) fields.Add("tech");
        return fields;
    }

    private static List<string> MatchProjectFields(ResumeProject p, string query)
    {
        var fields = new List<string>();
        if (ContainsIgnoreCase(p.Name, query)) fields.Add("name");
        if (ContainsIgnoreCase(p.Summary, query)) fields.Add("summary");
        if (p.Outcomes.Any(o => ContainsIgnoreCase(o, query))) fields.Add("outcomes");
        if (p.Tech.Any(t => ContainsIgnoreCase(t, query))) fields.Add("tech");
        return fields;
    }

    private static List<string> MatchNarrativeFields(ResumeNarrative n, string query)
    {
        var fields = new List<string>();
        if (ContainsIgnoreCase(n.OriginStory, query)) fields.Add("originStory");
        if (ContainsIgnoreCase(n.Bridge, query)) fields.Add("bridge");
        if (ContainsIgnoreCase(n.Carryover, query)) fields.Add("carryover");
        return fields;
    }

    private static List<string> MatchFaqFields(ResumeFaqEntry f, string query)
    {
        var fields = new List<string>();
        if (ContainsIgnoreCase(f.Question, query)) fields.Add("question");
        if (ContainsIgnoreCase(f.Answer, query)) fields.Add("answer");
        return fields;
    }

    private static List<string> MatchPersonFields(ResumePerson p, string query)
    {
        var fields = new List<string>();
        if (ContainsIgnoreCase(p.Summary, query)) fields.Add("summary");
        if (ContainsIgnoreCase(p.LookingFor, query)) fields.Add("lookingFor");
        if (ContainsIgnoreCase(p.Availability, query)) fields.Add("availability");
        if (ContainsIgnoreCase(p.WorkAuth, query)) fields.Add("workAuth");
        if (ContainsIgnoreCase(p.TimeZone, query)) fields.Add("timeZone");
        if (ContainsIgnoreCase(p.Compensation, query)) fields.Add("compensation");
        if (ContainsIgnoreCase(p.PortfolioSite, query)) fields.Add("portfolioSite");
        if (ContainsIgnoreCase(p.Email, query)) fields.Add("email");
        if (ContainsIgnoreCase(p.Github, query)) fields.Add("github");
        if (ContainsIgnoreCase(p.Linkedin, query)) fields.Add("linkedin");
        if (ContainsIgnoreCase(p.FreelanceSite, query)) fields.Add("freelanceSite");
        if (p.EmploymentTypes.Any(t => ContainsIgnoreCase(t, query))) fields.Add("employmentTypes");
        return fields;
    }

    private static bool ContainsIgnoreCase(string? haystack, string needle) =>
        !string.IsNullOrEmpty(haystack) && haystack.Contains(needle, StringComparison.OrdinalIgnoreCase);

    /// <summary>
    /// Matches a stack entry to the visitor's skill phrase: exact (case-insensitive), or substring when
    /// <paramref name="skill"/> is at least two characters so ".NET" hits stack labels that contain it (including "ASP.NET Core").
    /// </summary>
    private static bool TechEntryMatchesSkill(string tech, string skill)
    {
        if (tech.Equals(skill, StringComparison.OrdinalIgnoreCase))
            return true;
        if (skill.Length < 2)
            return false;
        return tech.Contains(skill, StringComparison.OrdinalIgnoreCase);
    }

    private JsonElement RunListProjectsBySkill(JsonElement input)
    {
        if (!ResumeToolInputValidator.TryDeserializeListProjectsBySkill(input, out var model, out var err) || model is null)
            return ValidationError(err ?? "deserialize_failed");

        if (string.IsNullOrWhiteSpace(model.Skill))
            return Error("'skill' is required.");

        var skill = model.Skill.Trim();
        var matches = _resume.Projects
            .Where(p => p.Tech.Any(t => TechEntryMatchesSkill(t, skill)))
            .Select(ProjectToObject)
            .ToList();

        return JsonSerializer.SerializeToElement(new { skill, items = matches, count = matches.Count });
    }

    private JsonElement RunGetMetrics(JsonElement input)
    {
        if (!ResumeToolInputValidator.TryDeserializeGetMetrics(input, out var model, out var err) || model is null)
            return ValidationError(err ?? "deserialize_failed");

        var id = model.Id;

        var items = (id is null
                ? _resume.Metrics
                : _resume.Metrics.Where(m => string.Equals(m.Id, id, StringComparison.OrdinalIgnoreCase)))
            .Select(m => (object)new
            {
                id = m.Id,
                label = m.Label,
                value = m.Value,
                unit = m.Unit,
                note = m.Note,
            })
            .ToList();

        return JsonSerializer.SerializeToElement(new { items, count = items.Count });
    }

    private JsonElement RunListRecentShipped(JsonElement input)
    {
        if (!ResumeToolInputValidator.TryDeserializeListRecentShipped(input, out var model, out var err) || model is null)
            return ValidationError(err ?? "deserialize_failed");

        var limit = model.Limit ?? 5;
        if (limit < 1) limit = 1;
        if (limit > 20) limit = 20;

        var items = _resume.Projects
            .Where(p => string.Equals(p.Status, "shipped", StringComparison.OrdinalIgnoreCase))
            .OrderByDescending(p => p.Year)
            .Take(limit)
            .Select(ProjectToObject)
            .ToList();

        return JsonSerializer.SerializeToElement(new { items, count = items.Count, limit });
    }

    private JsonElement RunGetNarrative()
    {
        var n = _resume.Narrative;
        return JsonSerializer.SerializeToElement(new
        {
            originStory = n.OriginStory,
            bridge = n.Bridge,
            carryover = n.Carryover,
        });
    }

    private JsonElement RunGetFaq(JsonElement input)
    {
        if (!ResumeToolInputValidator.TryDeserializeGetFaq(input, out var model, out var err) || model is null)
            return ValidationError(err ?? "deserialize_failed");

        IEnumerable<ResumeFaqEntry> source = _resume.Faq;

        if (!string.IsNullOrWhiteSpace(model.Id))
        {
            var id = model.Id.Trim();
            source = source.Where(f => string.Equals(f.Id, id, StringComparison.OrdinalIgnoreCase));
        }

        if (!string.IsNullOrWhiteSpace(model.Keyword))
        {
            var kw = model.Keyword.Trim();
            source = source.Where(f =>
                ContainsIgnoreCase(f.Question, kw) || ContainsIgnoreCase(f.Answer, kw));
        }

        var items = source
            .Select(f => (object)new { id = f.Id, question = f.Question, answer = f.Answer })
            .ToList();

        return JsonSerializer.SerializeToElement(new { items, count = items.Count });
    }

    private static object RoleToObject(ResumeRole r) => new
    {
        id = r.Id,
        title = r.Title,
        org = r.Org,
        startYear = r.StartYear,
        endYear = r.EndYear,
        remote = r.Remote,
        summary = r.Summary,
        achievements = r.Achievements,
        tech = r.Tech,
        url = r.Url,
    };

    private static object ProjectToObject(ResumeProject p) => new
    {
        id = p.Id,
        name = p.Name,
        roleId = p.RoleId,
        year = p.Year,
        status = p.Status,
        visibility = p.Visibility,
        url = p.Url,
        repoUrl = p.RepoUrl,
        demoUrl = p.DemoUrl,
        summary = p.Summary,
        outcomes = p.Outcomes,
        tech = p.Tech,
    };

    private static JsonElement ValidationError(string details) =>
        JsonSerializer.SerializeToElement(new { error = "validation_failed", details });

    private static JsonElement Error(string message) =>
        JsonSerializer.SerializeToElement(new { error = message });
}

File: mcp/src/resumeTools.ts
import type {
  ResumeData,
  ResumeFaqEntry,
  ResumeNarrative,
  ResumePerson,
  ResumeProject,
  ResumeRole,
} from "./resumeTypes.js";
import { ToolNames, type ToolName } from "./toolDefinitions.js";

function errorMessage(message: string): { error: string } {
  return { error: message };
}

function containsIgnoreCase(haystack: string | undefined | null, needle: string): boolean {
  if (haystack === undefined || haystack === null || haystack === "") return false;
  return haystack.toLowerCase().includes(needle.toLowerCase());
}

function techEntryMatchesSkill(tech: string, skill: string): boolean {
  if (tech.toLowerCase() === skill.toLowerCase()) return true;
  if (skill.length < 2) return false;
  return tech.toLowerCase().includes(skill.toLowerCase());
}

function matchesRole(
  role: ResumeRole,
  id: string | undefined,
  org: string | undefined,
  year: number | undefined,
  currentYear: number
): boolean {
  if (id !== undefined && id !== null && role.id.toLowerCase() !== id.toLowerCase()) return false;
  if (org !== undefined && org !== null && !(role.org ?? "").toLowerCase().includes(org.toLowerCase()))
    return false;
  if (year !== undefined && year !== null) {
    const endYear = role.endYear ?? currentYear;
    if (year < role.startYear || year > endYear) return false;
  }
  return true;
}

function roleToObject(r: ResumeRole) {
  return {
    id: r.id,
    title: r.title ?? "",
    org: r.org ?? "",
    startYear: r.startYear,
    endYear: r.endYear ?? null,
    remote: r.remote ?? false,
    summary: r.summary ?? "",
    achievements: r.achievements ?? [],
    tech: r.tech ?? [],
    url: r.url ?? null,
  };
}

function projectToObject(p: ResumeProject) {
  return {
    id: p.id,
    name: p.name ?? "",
    roleId: p.roleId ?? null,
    year: p.year,
    status: p.status ?? "",
    visibility: p.visibility ?? null,
    url: p.url ?? null,
    repoUrl: p.repoUrl ?? null,
    demoUrl: p.demoUrl ?? null,
    summary: p.summary ?? "",
    outcomes: p.outcomes ?? [],
    tech: p.tech ?? [],
  };
}

function matchRoleFields(r: ResumeRole, query: string): string[] {
  const fields: string[] = [];
  if (containsIgnoreCase(r.title, query)) fields.push("title");
  if (containsIgnoreCase(r.org, query)) fields.push("org");
  if (containsIgnoreCase(r.summary, query)) fields.push("summary");
  if ((r.achievements ?? []).some((a) => containsIgnoreCase(a, query))) fields.push("achievements");
  if ((r.tech ?? []).some((t) => containsIgnoreCase(t, query))) fields.push("tech");
  return fields;
}

function matchProjectFields(p: ResumeProject, query: string): string[] {
  const fields: string[] = [];
  if (containsIgnoreCase(p.name, query)) fields.push("name");
  if (containsIgnoreCase(p.summary, query)) fields.push("summary");
  if ((p.outcomes ?? []).some((o) => containsIgnoreCase(o, query))) fields.push("outcomes");
  if ((p.tech ?? []).some((t) => containsIgnoreCase(t, query))) fields.push("tech");
  return fields;
}

function matchNarrativeFields(n: ResumeNarrative, query: string): string[] {
  const fields: string[] = [];
  if (containsIgnoreCase(n.originStory, query)) fields.push("originStory");
  if (containsIgnoreCase(n.bridge, query)) fields.push("bridge");
  if (containsIgnoreCase(n.carryover, query)) fields.push("carryover");
  return fields;
}

function matchFaqFields(f: ResumeFaqEntry, query: string): string[] {
  const fields: string[] = [];
  if (containsIgnoreCase(f.question, query)) fields.push("question");
  if (containsIgnoreCase(f.answer, query)) fields.push("answer");
  return fields;
}

function matchPersonFields(p: ResumePerson, query: string): string[] {
  const fields: string[] = [];
  if (containsIgnoreCase(p.summary, query)) fields.push("summary");
  if (containsIgnoreCase(p.lookingFor, query)) fields.push("lookingFor");
  if (containsIgnoreCase(p.availability, query)) fields.push("availability");
  if (containsIgnoreCase(p.workAuth, query)) fields.push("workAuth");
  if (containsIgnoreCase(p.timeZone, query)) fields.push("timeZone");
  if (containsIgnoreCase(p.compensation, query)) fields.push("compensation");
  if (containsIgnoreCase(p.portfolioSite, query)) fields.push("portfolioSite");
  if (containsIgnoreCase(p.email, query)) fields.push("email");
  if (containsIgnoreCase(p.github, query)) fields.push("github");
  if (containsIgnoreCase(p.linkedin, query)) fields.push("linkedin");
  if (containsIgnoreCase(p.freelanceSite, query)) fields.push("freelanceSite");
  if ((p.employmentTypes ?? []).some((t) => containsIgnoreCase(t, query))) fields.push("employmentTypes");
  return fields;
}

function normalizeArgs(raw: unknown): Record<string, unknown> {
  if (raw === undefined || raw === null) return {};
  if (typeof raw === "object" && !Array.isArray(raw)) return raw as Record<string, unknown>;
  return {};
}

export function runResumeTool(
  resume: ResumeData,
  tool: ToolName,
  args: unknown,
  currentYearUtc: number
): unknown {
  const input = normalizeArgs(args);

  switch (tool) {
    case ToolNames.GetRole: {
      const id = input.id !== undefined && input.id !== null ? String(input.id) : undefined;
      const org = input.org !== undefined && input.org !== null ? String(input.org) : undefined;
      const year =
        input.year !== undefined && input.year !== null && typeof input.year === "number"
          ? input.year
          : undefined;
      const roles = resume.roles ?? [];
      const matches = roles
        .filter((r) => matchesRole(r, id, org, year, currentYearUtc))
        .map(roleToObject);
      return { items: matches, count: matches.length };
    }
    case ToolNames.SearchResume: {
      const queryRaw = input.query;
      if (typeof queryRaw !== "string" || queryRaw.trim() === "")
        return errorMessage("'query' is required.");
      const query = queryRaw.trim();
      const items: { kind: string; id: string; matchedFields: string[] }[] = [];
      for (const r of resume.roles ?? []) {
        const matched = matchRoleFields(r, query);
        if (matched.length > 0) items.push({ kind: "role", id: r.id, matchedFields: matched });
      }
      for (const p of resume.projects ?? []) {
        const matched = matchProjectFields(p, query);
        if (matched.length > 0) items.push({ kind: "project", id: p.id, matchedFields: matched });
      }
      const narrative = resume.narrative ?? { originStory: "", bridge: "", carryover: "" };
      const narrativeMatched = matchNarrativeFields(narrative, query);
      if (narrativeMatched.length > 0)
        items.push({ kind: "narrative", id: "narrative", matchedFields: narrativeMatched });
      for (const f of resume.faq ?? []) {
        const matched = matchFaqFields(f, query);
        if (matched.length > 0) items.push({ kind: "faq", id: f.id, matchedFields: matched });
      }
      const person = resume.person ?? {};
      const personMatched = matchPersonFields(person, query);
      if (personMatched.length > 0)
        items.push({ kind: "person", id: "person", matchedFields: personMatched });
      return { query, items, count: items.length };
    }
    case ToolNames.ListProjectsBySkill: {
      const skillRaw = input.skill;
      if (typeof skillRaw !== "string" || skillRaw.trim() === "")
        return errorMessage("'skill' is required.");
      const skill = skillRaw.trim();
      const matches = (resume.projects ?? [])
        .filter((p) => (p.tech ?? []).some((t) => techEntryMatchesSkill(t, skill)))
        .map(projectToObject);
      return { skill, items: matches, count: matches.length };
    }
    case ToolNames.GetMetrics: {
      const id =
        input.id !== undefined && input.id !== null && String(input.id).trim() !== ""
          ? String(input.id).trim()
          : undefined;
      const metrics = resume.metrics ?? [];
      const filtered =
        id === undefined
          ? metrics
          : metrics.filter((m) => m.id.toLowerCase() === id.toLowerCase());
      const items = filtered.map((m) => ({
        id: m.id,
        label: m.label ?? "",
        value: m.value,
        unit: m.unit ?? "",
        note: m.note ?? null,
      }));
      return { items, count: items.length };
    }
    case ToolNames.ListRecentShipped: {
      let limit = input.limit !== undefined && input.limit !== null ? Number(input.limit) : 5;
      if (Number.isNaN(limit)) limit = 5;
      if (limit < 1) limit = 1;
      if (limit > 20) limit = 20;
      const items = (resume.projects ?? [])
        .filter((p) => (p.status ?? "").toLowerCase() === "shipped")
        .sort((a, b) => b.year - a.year)
        .slice(0, limit)
        .map(projectToObject);
      return { items, count: items.length, limit };
    }
    case ToolNames.GetNarrative: {
      const n = resume.narrative ?? { originStory: "", bridge: "", carryover: "" };
      return {
        originStory: n.originStory ?? "",
        bridge: n.bridge ?? "",
        carryover: n.carryover ?? "",
      };
    }
    case ToolNames.GetFaq: {
      let source = [...(resume.faq ?? [])];
      const idRaw = input.id;
      if (typeof idRaw === "string" && idRaw.trim() !== "") {
        const id = idRaw.trim();
        source = source.filter((f) => f.id.toLowerCase() === id.toLowerCase());
      }
      const kwRaw = input.keyword;
      if (typeof kwRaw === "string" && kwRaw.trim() !== "") {
        const kw = kwRaw.trim();
        source = source.filter(
          (f) => containsIgnoreCase(f.question, kw) || containsIgnoreCase(f.answer, kw)
        );
      }
      const items = source.map((f) => ({
        id: f.id,
        question: f.question ?? "",
        answer: f.answer ?? "",
      }));
      return { items, count: items.length };
    }
    default:
      return errorMessage(`Unknown tool '${String(tool)}'.`);
  }
}
